### ISSUE TITLE
SWIG failure against Octave 6.0

### ISSUE THREAD

**Issue by rlaboiss** [swig/swig#1893]
A test tarball of the upcoming version 6.0 of Octave [has been announced](https://mail.gnu.org/archive/html/octave-maintainers/2020-08/msg00018.html). I tried the example in this [zip file](https://github.com/swig/swig/files/5493683/octave-swig-example.zip) against Octave 6.0.90, with the following commands:

```sh
swig -octave -c++ example.i 
mkoctfile example_wrap.cxx example.c > swig-mkoctfile.log 2>&1 
```

Compilation fails with loads of errors, as can be seen in the swig-mkoctfile.log file in the tarball.

I hope SWIG will be prepared for the upcoming release of Octave 6.

**Comment by rlaboiss** [comment]
I have [reported](https://savannah.gnu.org/bugs/?59422) this issue in the Octave's Bug Tracking System (even though it is not a bug in Octave).

Please, notice that Jonh Eaton, the lead Octave developer, has already [posted](https://savannah.gnu.org/bugs/?59422#comment1) a long comment on the issue.  I hope the SWIG developers will take a look at it.

**Comment by wsfulton** [comment]
@kwwete has contributed a lot to the SWIG Octave module but has now stepped back. Future Octave enhancements will need to come from Octave users of the SWIG module.

**Comment by dirteat** [comment]
Too bad :(
Hitting that bug in various other projects, like plplot.

**Comment by opoplawski** [comment]
octave 6.1 is out and I'm looking to update it in Fedora.  Would be nice to still have swig support.

**Comment by galtieri** [comment]
a real serious trouble :(

**Comment by kwwette** [comment]
I read through the report issue on the Octave bug tracker. It sounds like this shouldn't be too hard to fix. While the Octave API has deprecated some functions, it looks like they all have replacements or possible workarounds, so it should be possible to make these replacements (wrapped in `#if !SWIG_OCTAVE_PREREQ(...)` directives for backward compatibility). It would however need someone willing to dig into the Octave module and figure out where the errors are - unfortunately I no longer have the time to do that.

The wider problem with the SWIG Octave module is that Octave has no stable C API. Pretty much every new Octave release makes backward-incompatible changes to the API which need to be worked around. And because SWIG releases are obviously not synced up to Octave releases, more often than not the latest version of SWIG cannot work with the latest version of Octave. This situation is partly why I've moved away from using Octave with SWIG, to avoid this constant maintenance.

A long-term solution would be to drop the current SWIG Octave module entirely, and support both Octave and MATLAB through the latter's MEX interface, which Octave also supports, and which I hope is a more stable API. I have a dim memory of a MATLAB module being worked on for SWIG, but don't know the latest. I would imagine SWIG support for MATLAB (with Octave support as a bonus) would be attractive to many people, and you might be more likely to find willing maintainers with a wider user base.

**Comment by rwf1** [comment]
I have a proposed patch to answer this issue.  I am including in this a patch to deal with another issue that I have had do deal with in moving to Octave 6.0 although I have provided separate patches for these.

First I have a [Readme.txt](https://github.com/swig/swig/files/6400205/Readme.txt) that describes the patches.  

The first patch [SwigAddedExample.patch.txt](https://github.com/swig/swig/files/6400222/SwigAddedExample.patch.txt)  adds an example that illustrates a problem I ran into when compiling a wrapped file with -Bsymbolic in recent releases of Octave including 6.2. 

The second patch [SwigTolerateBsymbolic.patch.txt](https://github.com/swig/swig/files/6400229/SwigTolerateBsymbolic.patch.txt) is a proposed change to deal with this issue.

The third patch [SwigOctave-6.patch.txt](https://github.com/swig/swig/files/6400241/SwigOctave-6.patch.txt) is a propose patch that addresses the topic of this thread - that is letting swig work with Octave-6.2.

The fourth patch  [SwigOctave-6NameChanges.patch.txt](https://github.com/swig/swig/files/6400252/SwigOctave-6NameChanges.patch.txt)  has some added name changes that were not flagged in the Examples or tests but these kind of name changes came up in some code I am working on so when I found them in the swig interface code I corrected them.

Finally I made a composite of these patches [SwigOctave-6Composite.patch.txt](https://github.com/swig/swig/files/6400277/SwigOctave-6Composite.patch.txt)  in case that is easier to deal with.

I have run the Examples and tests in swig for the octave interface against Octave-4.2.2 built locally from a tar file, Octave-5.2 as shipped with Fedora-33, and Octave-6.2 build locally from a tar file.

**Comment by alatina** [comment]
I just tried this patch. It works great on my system, both with Octave 6.2 and 5.2. Thanks a lot, @rwf1 !

**Comment by alatina** [comment]
Since I am not a SWIG developer, I was wondering: could anyone apply the patch proposed by @rwf1 on master, so that it'll appear in the next SWIG release? Perhaps @ojwb, that has recently modified https://github.com/swig/swig/blob/master/Source/Modules/octave.cxx?

**Comment by ojwb** [comment]
I only touched that file as part of a global clean-up.  I don't know anything specifically about octave, so I'm not really the right person to review Octave-specific changes.  (And if global clean-ups meant I was responsible for reviewing any proposed changes to any file I last touched, I'd be highly discouraged from doing them...)

@rwf1 If you want to propose a patch, please open a pull-request for it - that way it gets automatically built and the testsuite run so reviewers can easily see if it introduces problems either for one of the tested Octave versions or for another language.

**Comment by rwf1** [comment]
I created a pull request   #2020

**Comment by wsfulton** [comment]
Thanks for all the comments and testing. All fixed now, see #2020.
